### DISCUSSION TITLE
`IntersectionObserver` via Ref working in development but not production

### DISCUSSION THREAD

**Discussion by fabiancdng** [vercel/next.js#48500] [Help] [answered]
### Summary

Hey 👋 
I was wondering if someone here might be able to help out with the following issue I'm running into:

I have a React context that holds a value for the active item in a navbar.

On the homepage, I want to update that state within the context using `IntersectionObserver` (when a user scrolls in or out a certain div with the ID "projects").

I create and subscribe the `IntersectionObserver` in a `useEffect` hook:

```
const projectsSectionRef = useRef<HTMLDivElement>(null);

  // Get values from global website context to set active nav item.
  const { setActiveNavItem } = useContext(GlobalsContext);

  useEffect(() => {
    // Create IntersectionObserver to change active nav item based on section in viewport.
    const observer = new IntersectionObserver(
      (entries) => {
        if (entries[0].isIntersecting) {
          setActiveNavItem('/#projects');
        } else {
          setActiveNavItem('/');
        }
      },
      { threshold: 0.2 }
    );

    // Observe whether #projects section is in viewport.
    observer.observe(projectsSectionRef.current!);

    return () => {
      // Disconnect observer when component unmounts.
      observer.disconnect();
    };
    // Disable exhaustive-deps warning because setter is not needed in dependency array.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [projectsSectionRef]);
```

The div:

```
<div id="projects" ref={projectsSectionRef}> ... </div>
```

Weirdly, when running this with `npm run dev` it works without any issues... But when running it in production, the state doesn't change (`IntersectionObserver` doesn't seem to call the callback function).

Does someone have any ideas what the problem could be here?

**Comment by icyJoseph** [comment]
Hi,

Does it work on dev if you turn off strict mode? 

You could try to log inside the clean up function:

```js
    return () => {
      // Disconnect observer when component unmounts.
      observer.disconnect();
    };
```

To make sure that the observer hasn't been disconnected. 

I am also a bit annoyed by the `eslint-disable-next-line`... just put the `setActiveNavItem` there, and the `ref` is not needed on that array.

Last but not least, how is this component being rendered? Are you in control of it using it as JSX?

**Reply by fabiancdng** [reply]
Hi @icyJoseph, thanks for your help! It works in dev with both strict mode on and off. When I console.log in the clean up function, nothing appears so the observer doesn't seem to be disconnecting. Even when I remove the clean up function entirely it still works in dev and doesn't in production.
It's a functional component that returns the `<div>` and its children as JSX.

**Reply by icyJoseph** [reply, marked as answer]
Alright, so no disconnecting, but what about the callback, is it being called at all?

Also could you make sure that `projectsSectionRef.current` is a thing? Do `console.log(projectsSectionRef.current)`, and see if it is an element, and if it is still present in the DOM. Maybe storyblok does something under the hood? 

While we are on that, if you remove the storyblok API, for testing only, and just let it have a div with some mocked children, does it work?

**Reply by fabiancdng** [reply]
@icyJoseph You got it! Actually this issue was related to the Storyblok API. Not sure what the issue was exactly but removing and re-inserting the component in the editor did the trick 🙈 Thank you a lot for your help!

**Comment by ggomaeng** [comment]
Seems to not work for me either 😭 It works in dev, but once I build and test it it doesn't seem to work properly.

**Reply by icyJoseph** [reply]
Hej, have you got a repository, demo to check this out?

**Reply by ggomaeng** [reply]
I'll try to whop it up tomorrow and share it here!

**Reply by anwesha-palit-tfs** [reply]
I am having the exact same issue with IntersectionObserver via ref. In prod build `ref.current` is null even if I can see the element in the DOM. Not sure how to get this to work. In dev env, the HTML element is correctly getting passed in ref

**Reply by icyJoseph** [reply]
Are you also using Storyblok? Got a repo/reproduction demo I could see?

**Reply by anwesha-palit-tfs** [reply]
Hi

Here is a demo of hook 

```
import {useState, useEffect} from "react"

const useOnScreen = (ref:any) => {
    const [isVisible, setIsVisible] = useState<Boolean>(false);

    const callback = (entries:any) => {
        const [entry] = entries
        console.log("OBSERVER TRIGGERED",entry.target,entry.isIntersecting)
        setIsVisible(entry.isIntersecting)
        return
    }
    
    const options = {
        threshold : 0.2,
        rootMargin : '-80px'
    }

    useEffect(() => {
        const observer = new IntersectionObserver(callback, options)
        if(typeof window !== 'undefined' && ref?.current){
            observer.observe(ref.current)
        }
        return () => {if(ref.current) observer.disconnect()}
    }, [isVisible])
     
    return isVisible;
}

export default useOnScreen
```


And I am using it in the app here 

```
function App(){

    const applicationRef = useRef<HTMLDivElement>()

    const applicationRefVal = useOnScreen(applicationRef)

    const [isApplicationVisible, setApplicationVisible] = useState<Boolean>(false)
    
    useEffect(()=>{
        if(!isApplicationVisible){
          setApplicationVisible(applicationRefVal)
          return
        }
      },[applicationRefVal])


    return(<div>
        <div ref={applicationRef}>{ isApplicationVisible && <dynamically loading a component> }</div>
    </div>)


}

```

For the dynamic loading I am using `dynamic/next`

**Reply by icyJoseph** [reply]
Try to format it using triple backticks: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by icyJoseph** [reply]
I copied your code, no changes, and it works:

<img width="1364" alt="Screenshot 2023-04-25 at 15 34 28" src="https://user-images.githubusercontent.com/21013447/234293703-ddd68115-066e-4004-bf5c-ca91d5e5812d.png">

**Reply by anwesha-palit-tfs** [reply]
It works for me too in dev, it does not work in production build. The ref and ref.current is undefined in the hook.

**Reply by icyJoseph** [reply]
Here's what I run (annotated your initial code):

```tsx
import Head from "next/head";
import { Inter } from "next/font/google";
import styles from "@/styles/Home.module.css";
import { useEffect, useRef, useState } from "react";
import useOnScreen from "@/use-on-screen";

const inter = Inter({ subsets: ["latin"] });

export default function Home() {
  const applicationRef = useRef<HTMLDivElement>(null);
  // this whole thing feels unnecessary, you just need isApplicationVisible, right?
  const applicationRefVal = useOnScreen(applicationRef);
  const [isApplicationVisible, setApplicationVisible] =
    useState<Boolean>(false);
 // you don't need this imho
  useEffect(() => {
    if (!isApplicationVisible) {
      setApplicationVisible(applicationRefVal);
      return;
    }
  }, [applicationRefVal]);

  return (
      <main className={`${styles.main} ${inter.className}`}>
        <div style={{ height: "100vh" }}></div>
        <div ref={applicationRef}>Hello</div>
        <div style={{ height: "100vh" }}></div>
      </main>
  );
}
```

And the hooks (with annotations):

```tsx
import { useState, useEffect } from "react";

// options should be defined here

const useOnScreen = (ref: any) => {
  const [isVisible, setIsVisible] = useState(false);

  const callback = (entries: any) => {
    const [entry] = entries;
    console.log("OBSERVER TRIGGERED", entry.target, entry.isIntersecting);
    setIsVisible(entry.isIntersecting);
    return;
  };

  // this is allocated every time this hook runs, you should put it at the top of the module
  const options = {
    threshold: 0.2,
    rootMargin: "-80px",
  };

  useEffect(() => {
    // callback should just be defined inside this use-effect, otherwise you might encounter funny bugs
    const observer = new IntersectionObserver(callback, options);
    // inside a use-effect, you can expect window to be defined no need to check
    if (typeof window !== "undefined" && ref?.current) {
      observer.observe(ref.current);
    }
    return () => {
      if (ref.current) observer.disconnect();
    };
    // probably a good idea to include ref, and remove isVisible
  }, [isVisible]);

  return isVisible;
};

export default useOnScreen;
```

**Reply by icyJoseph** [reply]
@anwesha-palit-tfs it works in build mode for me as well. 

I've made some annotations to your hook though, maybe that fixes it in your case?

It should look, something like this:

```tsx
import { useState, useEffect } from "react";

const options = {
  threshold: 0.2,
  rootMargin: "-80px",
};

const useOnScreen = (ref: any) => {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    if (!ref.current) return;

    const callback: IntersectionObserverCallback = (entries) => {
      const [entry] = entries;
      console.log(
        "OBSERVER TRIGGERED",
        process.env.NODE_ENV,
        entry.target,
        entry.isIntersecting
      );
      setIsVisible(entry.isIntersecting);
    };

    const observer = new IntersectionObserver(callback, options);

    observer.observe(ref.current);

    return () => {
      observer.disconnect();
    };
  }, [ref]);

  return isVisible;
};

export default useOnScreen;
```

**Reply by anwesha-palit-tfs** [reply]
I updated the hook according to your comments, I didn't understand why we don't need the `useEffect` in App, I am storing the value of the flag in a global state so that the with every `isIntersection` the flag doesn't switch. 

But even with your mentioned changes, the prod build does not seem to work, any suggestions on how to make sure the `ref` is passed in the first render of the page ? Or could it be that the dynamic import in `nextjs` is not working as it should in prod build ?

**Reply by icyJoseph** [reply]
Aha, dynamic import? Missed that part... So the element you want to observe is being dynamically imported?

**Reply by anwesha-palit-tfs** [reply]
Yes correct. `const SomeApp = dynamic(() => import('./SomeApp'));`

`<div ref={applicationRef}>{ isApplicationVisible && <SomeApp/> }</div>`

**Reply by icyJoseph** [reply]
But you are not passing the ref into the dynamic component, the div should be visible. This works in prod for me:

```tsx
import { Inter } from "next/font/google";
import styles from "@/styles/Home.module.css";
import { useRef } from "react";
import useOnScreen from "@/use-on-screen";
import dynamic from "next/dynamic";

const DynHello = dynamic(() => import("@/Hello"));
const inter = Inter({ subsets: ["latin"] });

export default function Home() {
  const applicationRef = useRef<HTMLDivElement>(null);
  const isAppVisible = useOnScreen(applicationRef);

  return (
    <main className={`${styles.main} ${inter.className}`}>
      <div style={{ height: "100vh" }}></div>
      <div ref={applicationRef}>{isAppVisible && <DynHello />}</div>
      <div style={{ height: "100vh" }}></div>
    </main>
  );
}
```

**Reply by anwesha-palit-tfs** [reply]
It is still not working for me, can SSG be any cause ? Not sure how to debug this.

**Reply by icyJoseph** [reply]
Do you think you could setup a repository that fails? or if its possible, invite me to the project? check with your business if that's ok.

You could do, for testing purposes:

- Change `use-on-screen` to:
```tsx
import { useState, useEffect } from "react";

const options = {
  threshold: 0.2,
  rootMargin: "-80px",
};

const useOnScreen = (ref: any) => {
  const [isVisible, setIsVisible] = useState(false);

  useEffect(() => {
    const element = ref instanceof HTMLElement ? ref : ref?.current;

    if (!element) return;

    const callback: IntersectionObserverCallback = (entries) => {
      const [entry] = entries;
      console.log(
        "OBSERVER TRIGGERED",
        process.env.NODE_ENV,
        entry.target,
        entry.isIntersecting
      );
      setIsVisible(entry.isIntersecting);
    };

    const observer = new IntersectionObserver(callback, options);

    observer.observe(element);

    return () => {
      observer.disconnect();
    };
  }, [ref]);

  return isVisible;
};

export default useOnScreen;

```

And then do a hacky thing... `useState` instead of `useRef`:

```tsx
import { Inter } from "next/font/google";
import styles from "@/styles/Home.module.css";
import { useRef, useState } from "react";
import useOnScreen from "@/use-on-screen";
import dynamic from "next/dynamic";

const DynHello = dynamic(() => import("@/Hello"));
const inter = Inter({ subsets: ["latin"] });

export default function Home() {
  const [applicationRef, setRef] = useState<HTMLDivElement | null>(null);
  const isAppVisible = useOnScreen(applicationRef);

  console.log({ applicationRef, isAppVisible });

  return (
    <main className={`${styles.main} ${inter.className}`}>
      <div style={{ height: "100vh" }}></div>
      <div ref={setRef}>{isAppVisible && <DynHello />}</div>
      <div style={{ height: "100vh" }}></div>
    </main>
  );
}
```

**Reply by anwesha-palit-tfs** [reply]
I finally got it to work!

`ref` on first render was getting set to `null`, and therefore `ref.current` was always pointing to `null`. Your hacky way helped. Thank you. 

```
  const [applicationRef, setApplicationRef] = useState()

  const handleRef1 = useCallback((node)=>setApplicationRef(node),[])

  const applicationRefVal = useOnScreen(applicationRef)

  ...
  return <div ref={handleRef1}>{applicationRefVal && <SomeApp/>}</div>

```
The hook also needs to be updated, as now the observer should observe `ref` instead of `ref.current`

**Reply by icyJoseph** [reply]
That just tells me that you got something wrong going on, in your app, and this is but a way to make it obvious. You gotta find the actual problem and fix it.

Also you don't need to do:

```js
  const handleRef1 = useCallback((node)=>setApplicationRef(node),[])
```

Because the state setter is already a stable reference. You are just wasting memory there.
